sql/stats: refactor concurrency checks to use the new locks table

yuzefovich opened the pull request:
This commit changes how we enforce concurrency limits on auto stats jobs
from using the system.jobs table to using the newly-introduced
system.table_statistics_locks table. This allows us to have more
flexibility in concurrency control while reducing the pressure on the
jobs system.

Before this patch the behavior was the following:
- an attempt to start an auto full stat job would be denied if there is at
least one other stat job (excluding auto partial)
- an attempt to start an auto partial stat job would be denied if there
is at least one other stat job (excluding auto partial) OR there is
another auto partial on the target table.

Notably, running manual full and manual partial stats jobs would prohibit
starting auto stats jobs.

After this patch the behavior becomes:
- we make auto stats jobs completely independent from manual ones (i.e.
regardless of the number and type of manual stats jobs, we might start
auto stats jobs)
- for auto full:
  - at most one auto full job can run on a single table
  - globally, at most `sql.stats.automatic_full_concurrency_limit`
(currently 1 by default) auto full stats jobs can run on a cluster.
(The default value matches the previous behavior, but we'll increase it
soon.)
- for auto partial USING EXTREMES:
  - at most one auto full or auto partial USING extremes can run on
 a single table
  - globally, at most `sql.stats.automatic_extremes_concurrency_limit`
(currently 128 by default) auto partial USING EXTREMES jobs can run on
a cluster. (Previously, there was no limit here.)

Fixes: #155498

Release note (sql change): New cluster settings that control the number
of concurrent automatic statistic collection jobs have been introduced:
- `sql.stats.automatic_full_concurrency_limit` controls the number of
auto full statistics collections, 1 by default, matching the behavior
since introduction of automatic stats collection around v19.2.
- `sql.stats.automatic_extremes_concurrency_limit` controls the number
of auto partial USING EXTREMES statistics collections, 128 by default.
Note that at most one statistics collection can run on a single table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/158835)
<!-- Reviewable:end -->

yuzefovich committed:
sql/stats: refactor concurrency checks to use the new locks table

This commit changes how we enforce concurrency limits on auto stats jobs
from using the system.jobs table to using the newly-introduced
system.table_statistics_locks table. This allows us to have more
flexibility in concurrency control while reducing the pressure on the
jobs system.

Before this patch the behavior was the following:
- an attempt to start an auto full stat job would be denied if there is at
least one other stat job (excluding auto partial)
- an attempt to start an auto partial stat job would be denied if there
is at least one other stat job (excluding auto partial) OR there is
another auto partial on the target table.

Notably, running manual full and manual partial stats jobs would prohibit
starting auto stats jobs.

After this patch the behavior becomes:
- we make auto stats jobs completely independent from manual ones (i.e.
regardless of the number and type of manual stats jobs, we might start
auto stats jobs)
- for auto full:
  - at most one auto full job can run on a single table
  - globally, at most `sql.stats.automatic_full_concurrency_limit`
(currently 1 by default) auto full stats jobs can run on a cluster.
(The default value matches the previous behavior, but we'll increase it
soon.)
- for auto partial USING EXTREMES:
  - at most one auto full or auto partial USING extremes can run on
 a single table
  - globally, at most `sql.stats.automatic_extremes_concurrency_limit`
(currently 128 by default) auto partial USING EXTREMES jobs can run on
a cluster. (Previously, there was no limit here.)

Release note (sql change): New cluster settings that control the number
of concurrent automatic statistic collection jobs have been introduced:
- `sql.stats.automatic_full_concurrency_limit` controls the number of
auto full statistics collections, 1 by default, matching the behavior
since introduction of automatic stats collection around v19.2.
- `sql.stats.automatic_extremes_concurrency_limit` controls the number
of auto partial USING EXTREMES statistics collections, 128 by default.
Note that at most one statistics collection can run on a single table.

yuzefovich commented:
I think the optimization in the second commit is simple enough to include in the patch, but I extracted it out temporarily for ease of review. Additionally, I kept the behavior of allowing at most one auto full stat collection - I plan to do some QA later this week and want to increase the default for 26.1, but that will be in a separate patch.

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/create_stats.go` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-Og4gc9D2tYd-bwrDyGL:-Og4gc9D2tYd-bwrDyGM:bp31x19) ([raw file](https://github.com/cockroachdb/cockroach/blob/81fb6929e0cd162630f872abce6a6e728b476888/pkg/sql/create_stats.go#L287)):*
> ```Go
> // checkStatsJobsStillActive takes an array of job IDs and checks whether each
> // job is still active, returning a list of still active jobs.
> func checkStatsJobsStillActive(
> ```

I don't think this needs to be a separate query. Could we use something like this?

```sql
SELECT array_agg(id)
FROM (
  SELECT unnest(job_ids) AS job_id
  FROM system.table_statistics_locks
  WHERE table_id = $1
  AND kind = $2
  FOR UPDATE OF table_statistics_locks
)
INNER LOOKUP JOIN system.jobs ON id = job_id
WHERE status IN ('running', ...);
```


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/create_stats.go` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-Og4gc9D2tYd-bwrDyGL:-Og5BKCW4bVFkvFocg4f:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/81fb6929e0cd162630f872abce6a6e728b476888/pkg/sql/create_stats.go#L287)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I don't think this needs to be a separate query. Could we use something like this?

```sql
SELECT array_agg(id)
FROM (
  SELECT unnest(job_ids) AS job_id
  FROM system.table_statistics_locks
  WHERE table_id = $1
  AND kind = $2
  FOR UPDATE OF table_statistics_locks
)
INNER LOOKUP JOIN system.jobs ON id = job_id
WHERE status IN ('running', ...);
```
</blockquote></details>

Nice!


<!-- Sent from Reviewable.io -->

yuzefovich commented on pkg/sql/create_stats.go:
```diff
@@ -219,6 +257,120 @@ func (n *createStatsNode) runJob(ctx context.Context) error {
+  AND kind %s
+  FOR UPDATE OF table_statistics_locks
+)
+INNER LOOKUP JOIN system.jobs ON id = job_id
```
We had a concern in the unhappy case when a node crashes, meaning that if the stats job was responsible for updating the status in the locks table, what would happen if the coordinator node crashed? I'm guessing that the jobs system has recovery mechanism built-in (does it?), so we decided we'd piggy back on that. An alternative that Michael and I briefly discussed was introducing liveness-like behavior to the locks table for each running stats job - a stats job would be responsible for heartbeating its job_id in the locks table periodically so that if the coordinator node crashes, the job_id can expire some time soon.
<!-- Reviewable comment -Og8bmt1BYhAj0ggA0mz:bqarp25 -->
<!-- Sent from Reviewable.io -->

dt replied:
> the jobs system has recovery mechanism built-in

Indeed, a new node will adopt and run the job, at which point I think you get to assume that node will update the state you're looking at when it runs it, so I think you can look at that state instead of looking at the jobs system state?

yuzefovich replied:
(For some reason I thought that if the coordinator node crashed, we'd just give up on executing a particular job, but now that I spelled it out, it doesn't make much sense.)

yuzefovich replied:
Refactored the approach based on this discussion which eliminated the join against `system.jobs`.

Since for USING EXTREMES we need to aggregate across two rows, and aggregates aren't friends with SFU locking, I extracted the locking read into a subquery.
<!-- Reviewable comment -Ogjp4We2DqoyUjt2VI4:bvgtgub -->
<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @michae2)



<!-- Sent from Reviewable.io -->

yuzefovich commented on pkg/sql/create_stats.go:
```diff
@@ -219,6 +257,120 @@ func (n *createStatsNode) runJob(ctx context.Context) error {
 	return err
 }
 
+var checkFullStatsJobsQuery string
```
No strong preference from me so used your suggestion.
<!-- Reviewable comment -OgjoqDC1acU6aqmcV2R:by0f6s -->
<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
RFAL. I'm hoping we could get this backported to 26.1 and _perhaps_ increase the auto full global concurrency limit on the cloud later.

> @yuzefovich made 3 comments and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-:-OgjqXvh141rO8lZwnfz:b32580h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @msbutler).



<!-- Sent from Reviewable.io -->

dt commented:
> backport

I'm a little wary of a backport, at least with the big bump up to 128 concurrent jobs: We already see that stats jobs are _far and away_ the most numerous jobs created/processed in many clusters, and while we've taken some steps to react to that -- like auto-cleanup of succeeded jobs -- those steps were in response to the rate at which they were being created without concurrency = 1. A major change in that concurrency might change the situation here and feels like something I'd want to observe over longer development cycle than a backport after branch cut?

To clarify: no objection to backporting usage of the locks table to 26.1 (and indeed, would be happy to see it); my concern is specifically about increase to 128 concurrent jobs and its potential to massively increase stats jobs/day.

dt commented:
>  this PR actually adds a concurrency limit 

ah, got it, sounds great then

michae2 reviewed (commented):
This looks great! Just a couple suggestions.

> @michae2 reviewed 8 files and all commit messages, and made 5 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-:-OgnSH3u4p2mwx0EPHES:b-yi1ay9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, and @yuzefovich).

___
*[`pkg/sql/opt/exec/execbuilder/testdata/system` line 181 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-OgnRyF15QRg4IJBzRF-:-OgnRyF15QRg4IJBzRF0:byuoucg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b070ec934297a0c491e7f79161d96b8f2e774d91/pkg/sql/opt/exec/execbuilder/testdata/system#L181)):*
> ```
> │         label: buffer 1 (locked)
> │
> └── • subquery
> ```

I learned the other day that if you want to avoid the CTE (and the subquery) you can use `FOR UPDATE OF table_statistics_locks` like this:

```sql
SELECT array_cat_agg(job_ids)
FROM system.table_statistics_locks
WHERE table_id = 0
AND kind = 1
FOR UPDATE OF table_statistics_locks
```
___
*[`pkg/sql/create_stats.go` line 265 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-OgnOlN2CfYInct8aO1x:-OgnOlN2CfYInct8aO1y:bdxmu1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b070ec934297a0c491e7f79161d96b8f2e774d91/pkg/sql/create_stats.go#L265)):*
> ```Go
> 	case jobspb.AutoStatsName:
> 		return 1
> 	case jobspb.AutoPartialStatsName:
> ```

I forget, are WHERE automatic partial stats going to use this same name, or have a different name? If they'll have the same name, then it seems like this limit will apply to both WHERE and USING EXTREMES auto partial stats.

If this limit will apply to both kinds of auto partial stats, maybe we should call it something like `sql.stats.automatic_partial_concurrency_limit` instead of `sql.stats.automatic_extremes_concurrency_limit`?
___
*[`pkg/sql/create_stats.go` line 390 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-OgnKjLQ7jP5Ufphhxxb:-OgnKY1Z5iut3kDhjCzH:b-3inetn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b070ec934297a0c491e7f79161d96b8f2e774d91/pkg/sql/create_stats.go#L390)):*
> ```Go
> 	_, err := txn.ExecEx(
> 		ctx, "create-stats-release-table-lock", txn.KV(), sessiondata.InternalExecutorOverride{},
> 		"DELETE FROM system.table_statistics_locks WHERE table_id = $1 AND kind = $2",
> ```

Instead of DELETE, WDYT of upserting an empty array here, so that the SFU of subsequent checks still has a row to lock?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

> @yuzefovich made 4 comments.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-:-OgnY3p0-V1oBs4OIzpY:bz5mair)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @michae2, and @msbutler).

___
*[`pkg/sql/create_stats.go` line 265 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-OgnOlN2CfYInct8aO1x:-OgnX0BZDcdLfIwyQ9Ko:b-6qg6ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/b070ec934297a0c491e7f79161d96b8f2e774d91/pkg/sql/create_stats.go#L265)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I forget, are WHERE automatic partial stats going to use this same name, or have a different name? If they'll have the same name, then it seems like this limit will apply to both WHERE and USING EXTREMES auto partial stats.

If this limit will apply to both kinds of auto partial stats, maybe we should call it something like `sql.stats.automatic_partial_concurrency_limit` instead of `sql.stats.automatic_extremes_concurrency_limit`?
</blockquote></details>

The existing WHERE partial stats aren't triggered automatically, but they do use the same name. The auto "fix misestimates" partial stats that we want for #152995 might use a different name (tbd), but they will definitely use a different "stats kind" and will have a separate cluster setting / concurrency limit. The details are still a bit unclear, but I'll make sure that "fix misestimates" are independent of USING EXTREMES.
___
*[`pkg/sql/create_stats.go` line 390 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-OgnKjLQ7jP5Ufphhxxb:-OgnWQjpAhV1DxS9ThZm:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b070ec934297a0c491e7f79161d96b8f2e774d91/pkg/sql/create_stats.go#L390)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Instead of DELETE, WDYT of upserting an empty array here, so that the SFU of subsequent checks still has a row to lock?
</blockquote></details>

Good point, done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/system` line 181 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-OgnRyF15QRg4IJBzRF-:-OgnWuDO6uMHV1r0qbZC:b-wdj8zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b070ec934297a0c491e7f79161d96b8f2e774d91/pkg/sql/opt/exec/execbuilder/testdata/system#L181)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I learned the other day that if you want to avoid the CTE (and the subquery) you can use `FOR UPDATE OF table_statistics_locks` like this:

```sql
SELECT array_cat_agg(job_ids)
FROM system.table_statistics_locks
WHERE table_id = 0
AND kind = 1
FOR UPDATE OF table_statistics_locks
```
</blockquote></details>

Nice, thanks!


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Awesome work!

> @michae2 reviewed 2 files and all commit messages, made 1 comment, and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158835#-:-OgnaYXD2efqLlVT9ybQ:b-vfu9do)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt and @msbutler).



<!-- Sent from Reviewable.io -->
